### PR TITLE
add auto augmentation

### PR DESCRIPTION
**Pull request by GT9505**
add auto augmentation in data pipeline

### CODE REVIEW

File: mmdet/datasets/pipelines/auto_augment.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+        return img, gt_bboxes, gt_masks, gt_seg
+
+    def __repr__(self):
+        return f'{self.__class__.__name__}' \
+            f'(auto_augment_policies={self.auto_augment_policies}'
+
+
+def swap_box(bboxes):
+    '''Swap bounding bboxes from x1y1x2y2(y1x1y2x2) order to
```
Comment (hellock) [inline mmdet/datasets/pipelines/auto_augment.py:130]: Use double quotes for docstrings.

File: mmdet/datasets/pipelines/auto_augment.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+from ..builder import PIPELINES
+
+
+@PIPELINES.register_module()
+class AutoAugment(object):
+    """Auto augmentation.
+
+    This data augmentation is proposed in
+    "Learning Data Augmentation Strategies for Object Detection".
+    Link: https://arxiv.org/pdf/1906.11172.pdf
```
Comment (hellock) [inline mmdet/datasets/pipelines/auto_augment.py:15]:
```
`Learning Data Augmentation Strategies for Object Detection <https://arxiv.org/abs/1906.11172>`_
```

File: mmdet/datasets/pipelines/auto_augment.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        >>>     ]
+        >>> ]
+        >>> augmentation = AutoAugment(auto_augment_policies)
+        >>> img = np.ones(100, 100, 3)
+        >>> gt_bboxes = np.ones(10, 4)
+        >>> results = dict(img=img, gt_bboxes=gt_bboxes)
+        >>> results = augmentation(results)
+    """
+
+    def __init__(self, auto_augment_policies=None):
```
Comment (hellock) [inline mmdet/datasets/pipelines/auto_augment.py:55]: `policies` is enough.

File: mmdet/datasets/pipelines/auto_augment.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        assert type(auto_augment_policies) is list and len(
```
Comment (hellock) [inline mmdet/datasets/pipelines/auto_augment.py:56]: assert isinstance(policies, list)

File: mmdet/datasets/pipelines/auto_augment.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+                'auto_augment_policies must be list and more than 0, ' \
+                'respectively.'
+            for augment in one_augment_policy:
+                assert type(augment) is dict and 'type' in augment, \
+                    'The type of each specifically augmentation must be dict' \
+                    "and contain the key of 'type'."
+
+        self.auto_augment_policies = auto_augment_policies.copy()
+        raise NotImplementedError('Auto augmentation is working in progress '
+                                  'and currently not callable.')
```
Comment (hellock) [inline mmdet/datasets/pipelines/auto_augment.py:73]: This error can be removed. As we did not provide configs before it is finished, most users will not use it.

File: mmdet/datasets/pipelines/auto_augment.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+            if 'gt_seg' in results:
+                results['gt_seg'] = gt_seg
+
+        return results
+
+    def auto_augment_core(self, img, gt_bboxes, gt_masks, gt_seg, policies):
+        h, w, c = img.shape
+        hw_rescale = np.array([[h, w, h, w]])
+        # from x1y1x2y2 order to y1x1y2x2 order
+        gt_bboxes = swap_box(gt_bboxes)
```
Comment (hellock) [inline mmdet/datasets/pipelines/auto_augment.py:99]: What's the benefits?

File: mmdet/datasets/pipelines/auto_augment.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        # normalize bboxes
+        gt_bboxes /= hw_rescale
+        gt_masks = copy.deepcopy(gt_masks)
+        gt_masks = getattr(gt_masks, 'masks', None)
```
Comment (hellock) [inline mmdet/datasets/pipelines/auto_augment.py:103]: What's the purpose of the above two lines

File: mmdet/datasets/pipelines/auto_augment.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+            for augment in policy:
+                assert isinstance(augment, dict) and 'type' in augment, \
+                    'The type of each specifically augmentation must be' \
+                    'dict and contain the key of "type".'
+
+        self.policies = copy.deepcopy(policies)
+
+    def __call__(self, results):
+        if self.policies is not None:
+            img = results['img']
```
Comment (ZwwWayne) [inline mmdet/datasets/pipelines/auto_augment.py:71]: We may use `img_fields` for more general usage.

File: mmdet/datasets/pipelines/auto_augment.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    """Auto augmentation.
+
+    This data augmentation is proposed in
+    `Learning Data Augmentation Strategies for Object Detection <https://arxiv.org/pdf/1906.11172>`_  # noqa: E501
+    Required key is "img", and optional keys are "gt_bboxes", "gt_masks",
+    "gt_semantic_seg".
+
+    Args:
+        policies (list[list[dict]]): The policies of auto augmentation. Each
+            element in policies is a specifically augmentation policy, and is
```
Comment (hellock) [inline mmdet/datasets/pipelines/auto_augment.py:20]: specifically -> specific

File: mmdet/datasets/pipelines/auto_augment.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+            composed by several augmentations (dict). When call AutoAugment, a
```
Comment (hellock) [inline mmdet/datasets/pipelines/auto_augment.py:21]: `When calling AutoAugment` or `When AutoAugment is called`

File: mmdet/datasets/pipelines/auto_augment.py [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        >>> img = np.ones(100, 100, 3)
+        >>> gt_bboxes = np.ones(10, 4)
+        >>> results = dict(img=img, gt_bboxes=gt_bboxes)
+        >>> results = augmentation(results)
+    """
+
+    def __init__(self, policies):
+        assert isinstance(policies, list) and len(policies) > 0, \
+            'The type and length of policies must be list and more than 0, ' \
+            'respectively.'
```
Comment (hellock) [inline mmdet/datasets/pipelines/auto_augment.py:56]: policies must be a non-empty list

File: mmdet/datasets/pipelines/auto_augment.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+                'list and more than 0, respectively.'
+            for augment in policy:
+                assert isinstance(augment, dict) and 'type' in augment, \
+                    'The type of each specifically augmentation must be' \
+                    'dict and contain the key of "type".'
+
+        self.policies = copy.deepcopy(policies)
+        transforms = []
+        for policy in self.policies:
+            transforms.append(Compose(policy))
```
Comment (hellock) [inline mmdet/datasets/pipelines/auto_augment.py:69]:
```python
transforms = [Compose(policy) for policy in self.policies]
```

File: mmdet/datasets/pipelines/auto_augment.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+        self.transforms = transforms
```
Comment (hellock) [inline mmdet/datasets/pipelines/auto_augment.py:70]:
```python
self.transforms = [Compose(policy) for policy in self.policies]
```

File: mmdet/datasets/pipelines/auto_augment.py [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+    def __call__(self, results):
+        transform = np.random.choice(self.transforms)
+        return transform(results)
+
+    def __repr__(self):
+        return f'{self.__class__.__name__}' \
+            f'(policies={self.policies}'
```
Comment (hellock) [inline mmdet/datasets/pipelines/auto_augment.py:78]: `return f'{self.__class__.__name__}(policies={self.policies}'`

**Comment by zxcvfd13502** [comment]
@GT9505 Did you tried this AutoAugment method on mmdetection? I cannot get the result reported in the paper(mAP 39.0) and I was wondering whether you could get it.

**Comment by rhythm-A** [comment]
Can you give me some advice to use auto_augment.py?such as polices：v1?

> @GT9505 Did you tried this AutoAugment method on mmdetection? I cannot get the result reported in the paper(mAP 39.0) and I was wondering whether you could get it.

**Pull request merged**